test/UnitTest/Microsoft.AspNetCore.OData.Test/Abstraction/ODataActionSelectorTestHelper.cs

```diff
@@ -0,0 +1,155 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Reflection;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+using Microsoft.AspNet.OData.Routing;
+using Microsoft.AspNet.OData.Extensions;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Mvc.Controllers;
+using Microsoft.AspNetCore.Mvc.Abstractions;
+using System.IO;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+using System.Linq;
+#if NETCOREAPP2_0
+using Microsoft.Extensions.DependencyInjection;
+    using Microsoft.Extensions.Logging;
+    using Microsoft.OData;
+    using Microsoft.AspNetCore.Mvc.Internal;
+#else
+using Moq;
+#endif
+
+
+namespace Microsoft.AspNet.OData.Test.Abstraction
+{
+    /// <summary>
+    /// Helper methods for working with ODataActionSelector tests
+    /// </summary>
+    public static class ODataActionSelectorTestHelper
+    {
+        /// <summary>
+        /// Checks whether the specified method is a suitable match
+        /// for the specified action
+        /// </summary>
+        /// <param name="action">Controller action to match</param>
+        /// <param name="method">Candidate method to compare</param>
+        /// <returns></returns>
+        public static bool ActionMatchesMethod(ActionDescriptor action, MethodInfo method)
+        {
+            if (action.DisplayName != method.Name)
+            {
+                return false;
+            }
+
+            var parameters = method.GetParameters();
+            if (parameters.Length != action.Parameters.Count)
+            {
+                return false;
+            }
+
+            for (var i = 0; i < parameters.Length; i++)
+            {
+                if (parameters[i].Name != action.Parameters[i].Name ||
+                    parameters[i].ParameterType != action.Parameters[i].ParameterType)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Creates an action selector, descriptors and route builder based on
+        /// the specified controller
+        /// </summary>
+        /// <param name="controllerType"></param>
+        /// <param name="actionName">Action name of the action descriptors to return</param>
+        /// <param name="routeBuilder"></param>
+        /// <param name="actionSelector"></param>
+        /// <param name="actionDescriptors"></param>
+        public static void SetupActionSelector(System.Type controllerType,
+           string actionName,
+           out IRouteBuilder routeBuilder,
+           out ODataActionSelector actionSelector,
+           out IReadOnlyList<ControllerActionDescriptor> actionDescriptors)
+#if NETCOREAPP2_0
```

Comment:
@habbes `routeBuilder` and `actionDescriptors` seems to be getting initialized the same way in both `#if` and `#else` block. Maybe you could move that logic above this line?

Reply:
Maybe even `serviceProvider` and `modelBinderFactory` - DRY ;)

Reply:
I thought it would be easier to read if I completely separated the netcore2.0 and netcore3.1 implementations, that way you only focus on the block that's relevant to the version you're viewing at any one time.

But if you believe that keeping the common part in one place and only separating the parts that are different is easier to read, then I can change it.